src/ring_theory/chain_of_divisors.lean

```diff
@@ -0,0 +1,246 @@
+  {c : fin (n + 1) → associates M} (h₁ : strict_mono c) :
+  ¬ is_unit (c i) :=
+dvd_not_unit.not_unit (associates.dvd_not_unit_iff_lt.2
+  (h₁ $ show (0 : fin (n + 1)) < i, from i.pos_iff_ne_zero.mpr i_pos))
+
+lemma first_of_chain_is_unit {q : associates M} {n : ℕ} {c : fin (n + 1) → associates M}
+  (h₁ : strict_mono c) (h₂ : ∀ {r}, r ≤ q ↔ ∃ i, r = c i) : is_unit (c 0) :=
+begin
+  obtain ⟨i, hr⟩ := h₂.mp associates.one_le,
+  rw [associates.is_unit_iff_eq_one, ← associates.le_one_iff, hr],
+  exact h₁.monotone (fin.zero_le i)
+end
+
+/-- The second element of a chain is irreducible. -/
+lemma second_of_chain_is_irreducible {q : associates M} {n : ℕ} (hn : n ≠ 0)
+  {c : fin (n + 1) → associates M} (h₁ : strict_mono c) (h₂ : ∀ {r}, r ≤ q ↔ ∃ i, r = c i)
+  (hq : q ≠ 0) : irreducible (c 1) :=
+begin
+  cases n, { contradiction },
+  refine (associates.is_atom_iff (ne_zero_of_dvd_ne_zero hq (h₂.2 ⟨1, rfl⟩))).mp ⟨_, λ b hb, _⟩,
+  { exact ne_bot_of_gt (h₁ (show (0 : fin (n + 2)) < 1, from fin.one_pos)) },
+  obtain ⟨⟨i, hi⟩, rfl⟩ := h₂.1 (hb.le.trans (h₂.2 ⟨1, rfl⟩)),
+  cases i,
+  { exact (associates.is_unit_iff_eq_one _).mp (first_of_chain_is_unit h₁ @h₂) },
+  { simpa [fin.lt_iff_coe_lt_coe] using h₁.lt_iff_lt.mp hb },
+end
+
+lemma prime_dvd_eq_second_of_chain {p q r : associates M} {n : ℕ} (hn : n ≠ 0)
+  {c : fin (n + 1) → associates M} (h₁ : strict_mono c)
+  (h₂ : ∀ {r : associates M}, r ≤ q ↔ ∃ i, r = c i) (hp : prime p) (hr : r ∣ q) (hp' : p ∣ r) :
+  p = c 1 :=
+begin
+  cases n,
+  { contradiction },
+  obtain ⟨i, rfl⟩ := h₂.1 (dvd_trans hp' hr),
+  refine congr_arg c (eq_of_ge_of_not_gt _ $ λ hi, _),
+  { rw [fin.le_iff_coe_le_coe, fin.coe_one, nat.succ_le_iff, ← fin.coe_zero,
+        ← fin.lt_iff_coe_lt_coe, fin.pos_iff_ne_zero],
+    rintro rfl,
+    exact hp.not_unit (first_of_chain_is_unit h₁ @h₂) },
+  obtain (rfl | ⟨j, rfl⟩) := i.eq_zero_or_eq_succ,
+  { cases hi },
+  refine not_irreducible_of_not_unit_dvd_not_unit
+    (dvd_not_unit.not_unit (associates.dvd_not_unit_iff_lt.2
+    (h₁ (show (0 : fin (n + 2)) < j, from _)) )) _ hp.irreducible,
+  { simpa [← fin.succ_zero_eq_one, fin.succ_lt_succ_iff] using hi },
+  { refine associates.dvd_not_unit_iff_lt.2 (h₁ _),
+    simpa only [fin.coe_eq_cast_succ] using fin.lt_succ }
+end
+
+lemma card_subset_divisors_le_length_of_chain {q : associates M}
+  {n : ℕ} {c : fin (n + 1) → associates M} (h₂ : ∀ {r}, r ≤ q ↔ ∃ i, r = c i)
+  {m : finset (associates M)} (hm : ∀ r, r ∈ m → r ≤ q) : m.card ≤ n + 1 :=
+begin
+  classical,
+  have mem_image : ∀ (r : associates M), r ≤ q → r ∈ finset.univ.image c,
+  { intros r hr,
+    obtain ⟨i, hi⟩ := h₂.1 hr,
+    exact finset.mem_image.2 ⟨i, finset.mem_univ _, hi.symm⟩ },
+  rw ←finset.card_fin (n + 1),
+  exact (finset.card_le_of_subset $ λ x hx, mem_image x $ hm x hx).trans finset.card_image_le,
+end
+
+variables [unique_factorization_monoid M]
+
+lemma mem_chain_eq_pow_second_of_chain {q r : associates M} {n : ℕ} (hn : n ≠ 0)
```

Comment:
I am not sure about this name, since there is no `∈` in the statement.

Reply:
I've replaced `mem_chain` by `element_of_chain`. Does that sound better?